File: lib/weblog_authoring/repository.rb
# frozen_string_literal: true

require "pathname"
require "securerandom"

module WeblogAuthoring
  TOKYO_OFFSET = "+09:00"

  RepositorySnapshot = Struct.new(:pages, :problems, :redirects, keyword_init: true) do
    def initialize(pages: [], problems: [], redirects: [])
      super(pages: Array(pages).freeze, problems: Array(problems).freeze, redirects: Array(redirects).freeze)
      freeze
    end

    def with_redirect(old_route:, new_route:)
      redirect = Redirect.new(old_route:, new_route:)
      return self if redirects.include?(redirect)

      self.class.new(pages:, problems:, redirects: redirects + [redirect])
    end
  end

  class FileTransaction
    def initialize
      @writes = {}
      @deletes = {}
    end

    def write(path, content)
      resolved_path = Pathname(path)
      @writes[resolved_path] = content
      @deletes.delete(resolved_path)
    end

    def delete(path)
      resolved_path = Pathname(path)
      @deletes[resolved_path] = true unless @writes.key?(resolved_path)
    end

    def commit
      affected_paths = (@writes.keys + @deletes.keys).uniq
      original = affected_paths.each_with_object({}) do |path, values|
        values[path] = path.binread if path.exist?
      end
      created = affected_paths - original.keys

      begin
        @writes.each do |path, content|
          replace(path, content)
        end
        @deletes.each_key do |path|
          path.unlink if path.exist?
        end
      rescue StandardError => error
        rollback(original, created, error)
      end
    end

    private

    def rollback(original, created, original_error)
      rollback_errors = []

      original.each do |path, content|
        replace(path, content)
      rescue StandardError => error
        rollback_errors << error
      end

      created.each do |path|
        path.unlink if path.exist?
      rescue StandardError => error
        rollback_errors << error
      end

      if rollback_errors.empty?
        raise original_error
      end

      details = rollback_errors.map(&:message).join("; ")
      raise RuntimeError, "file transaction failed: #{original_error.message}; rollback failed: #{details}"
    end

    def replace(path, content)
      path.dirname.mkpath
      temporary = path.dirname.join(".#{path.basename}.#{SecureRandom.hex(8)}.tmp")

      File.open(temporary, File::WRONLY | File::CREAT | File::EXCL, 0o600) do |stream|
        stream.binmode
        stream.write(content)
        stream.flush
        stream.fsync
      end

      File.rename(temporary, path)
    ensure
      temporary.unlink if temporary&.exist?
    end
  end

  class ContentRepository
    attr_reader :content_dir, :database_path, :database

    def initialize(content_dir, database_path, clock)
      @content_dir = Pathname(content_dir)
      @database_path = Pathname(database_path)
      @clock = clock
      @database = AuthoringDatabase.new(@database_path)
    end

    def refresh
      pages = []
      problems = []
      page_ids = {}
      routes = {}

      if content_dir.exist?
        Pathname.glob(content_dir.join("*.md").to_s).sort.each do |path|
          parsed = parse_path(path, problems)
          next if parsed.nil?

          begin
            document = with_links_and_tokyo_time(validate_external_document(parsed))
          rescue ArgumentError => error
            problems << PageProblem.new(path:, detail: error.message)
            next
          end

          if page_ids.key?(document.id)
            problems << PageProblem.new(path:, detail: "duplicate page id: #{document.id}")
            next
          end
          if routes.key?(document.route)
            problems << PageProblem.new(path:, detail: "duplicate route: #{document.route}")
            next
          end

          page_ids[document.id] = true
          routes[document.route] = true
          pages << document
        end
      end

      snapshot = RepositorySnapshot.new(pages:, problems:)
      database.rebuild(snapshot)
      snapshot
    end

    def get_page(page_id)
      refresh.pages.find { |page| page.id == page_id }
    end

    def find_route(route)
      normalized = route.to_s.sub(%r{\A/+}, "").sub(%r{/+\z}, "")
      refresh.pages.find { |page| page.route == normalized }
    end

    def list_pages(query: "", status: nil, empty_only: false)
      normalized_query = query.to_s.downcase

      refresh.pages.select do |page|
        status_match = status.nil? || page.status == status
        empty_match = !empty_only || page.empty?
        query_match = normalized_query.empty? ||
          page.display_title.downcase.include?(normalized_query) ||
          page.route.downcase.include?(normalized_query)
        status_match && empty_match && query_match
      end
    end

    def save_draft(request)
      snapshot = refresh
      current = current_page(snapshot, request.page_id)
      document = if current.nil?
        new_document(snapshot, request)
      else
        validate_expected_update(current, request)
        replace_document(
          current,
          body: request.body,
          title: current.page_type == "date" ? request.title : nil,
          updated_at: now
        )
      end

      document = with_links_and_tokyo_time(document)
      transaction = FileTransaction.new
      transaction.write(document.path, WeblogAuthoring.serialize_document(document))

      existing_names = snapshot.pages.each_with_object({}) do |page, names|
        names[page.name] = true unless page.name.nil?
      end

      document.links.each do |link|
        name = WeblogAuthoring.validate_page_name(link.name)
        next if existing_names.key?(name)

        empty_page = new_empty_named_page(snapshot, name)
        transaction.write(empty_page.path, WeblogAuthoring.serialize_document(empty_page))
        existing_names[name] = true
      end

      transaction.commit
      refresh
      get_page(document.id) || document
    end

    def rename_named_page(page_id, new_name)
      snapshot = refresh
      page = current_page(snapshot, page_id)
      raise ConflictError, "page does not exist: #{page_id}" if page.nil?
      raise ConflictError, "only named pages can be renamed" unless page.page_type == "named"

      normalized_name = WeblogAuthoring.validate_page_name(new_name)
      return page if normalized_name == page.name

      destination = WeblogAuthoring.page_path(content_dir, "named", name: normalized_name, page_date: nil)
      if snapshot.pages.any? { |other| other.id != page.id && other.name == normalized_name } || destination.exist?
        raise ConflictError, "page name already exists: #{normalized_name}"
      end

      changed_at = now
      renamed_page = replace_document(page, name: normalized_name, path: destination, updated_at: changed_at)
      transaction = FileTransaction.new

      snapshot.pages.each do |source|
        if source.id == page.id
          transaction.write(destination, WeblogAuthoring.serialize_document(renamed_page))
          next
        end

        body = WeblogAuthoring.replace_wiki_links(source.body, old_name: page.name.to_s, new_name: normalized_name)
        next if body == source.body

        updated_page = replace_document(source, body:, updated_at: changed_at)
        transaction.write(updated_page.path, WeblogAuthoring.serialize_document(updated_page))
      end

      transaction.delete(page.path)
      transaction.commit
      refresh
      get_page(page.id) || renamed_page
    end

    private

    def parse_path(path, problems)
      text = path.read(encoding: "UTF-8")
      parsed = WeblogAuthoring.parse_document(path, text)
      if parsed.is_a?(PageProblem)
        problems << parsed
        return nil
      end

      parsed
    rescue ArgumentError, Encoding::InvalidByteSequenceError, Encoding::UndefinedConversionError => error
      problems << PageProblem.new(path:, detail: "document is not UTF-8: #{error.message}")
      nil
    end

    def current_page(snapshot, page_id)
      return nil if page_id.nil?

      snapshot.pages.find { |page| page.id == page_id }
    end

    def new_document(snapshot, request)
      timestamp = now

      case request.page_type
      when "date"
        raise ArgumentError, "date pages require page_date" unless request.page_date.instance_of?(Date)
        if snapshot.pages.any? { |page| page.page_date == request.page_date }
          raise ConflictError, "date page already exists: #{request.page_date.iso8601}"
        end

        path = WeblogAuthoring.page_path(content_dir, "date", name: nil, page_date: request.page_date)
        raise ConflictError, "source path already exists: #{path.basename}" if path.exist?

        PageDocument.new(
          id: SecureRandom.uuid.delete("-"),
          page_type: "date",
          name: nil,
          page_date: request.page_date,
          title: request.title,
          status: "draft",
          created_at: timestamp,
          updated_at: timestamp,
          published_at: nil,
          path:,
          body: request.body,
          links: []
        )
      when "named"
        name = WeblogAuthoring.validate_page_name(request.name.to_s)
        if snapshot.pages.any? { |page| page.name == name }
          raise ConflictError, "page name already exists: #{name}"
        end

        path = WeblogAuthoring.page_path(content_dir, "named", name:, page_date: nil)
        raise ConflictError, "source path already exists: #{path.basename}" if path.exist?

        PageDocument.new(
          id: SecureRandom.uuid.delete("-"),
          page_type: "named",
          name:,
          page_date: nil,
          title: nil,
          status: "draft",
          created_at: timestamp,
          updated_at: timestamp,
          published_at: nil,
          path:,
          body: request.body,
          links: []
        )
      else
        raise ArgumentError, "unknown page type: #{request.page_type}"
      end
    end

    def new_empty_named_page(snapshot, name)
      new_document(snapshot, SaveRequest.new(page_type: "named", name:, body: ""))
    end

    def validate_expected_update(page, request)
      return if request.expected_updated_at.nil? || request.expected_updated_at == page.updated_at

      raise ConflictError, "page was updated by another edit"
    end

    def now
      value = @clock.call
      raise ArgumentError, "clock must return a Time" unless value.is_a?(Time)

      value.getlocal(TOKYO_OFFSET)
    end

    def validate_external_document(document)
      case document.page_type
      when "named"
        raise ArgumentError, "named pages must not have page_date" unless document.page_date.nil?
        normalized_name = WeblogAuthoring.validate_page_name(document.name.to_s)
        raise ArgumentError, "named page name must be normalized" unless normalized_name == document.name
      when "date"
        raise ArgumentError, "date pages must not have name" unless document.name.nil?
      else
        raise ArgumentError, "unknown page type: #{document.page_type}"
      end

      expected_path = WeblogAuthoring.page_path(
        content_dir,
        document.page_type,
        name: document.name,
        page_date: document.page_date
      )
      raise ArgumentError, "document is not at canonical source path: #{expected_path.basename}" unless document.path == expected_path

      document
    end

    def with_links_and_tokyo_time(document)
      [document.created_at, document.updated_at, document.published_at].compact.each do |value|
        raise ArgumentError, "document timestamps must be aware" if value.utc_offset.nil?
      end

      links = WeblogAuthoring.extract_wiki_links(document.body)
      links.each do |link|
        WeblogAuthoring.validate_page_name(link.name)
      end

      replace_document(
        document,
        created_at: document.created_at.getlocal(TOKYO_OFFSET),
        updated_at: document.updated_at.getlocal(TOKYO_OFFSET),
        published_at: document.published_at&.getlocal(TOKYO_OFFSET),
        links:
      )
    end

    def replace_document(document, **changes)
      PageDocument.new(**document.to_h.merge(changes))
    end
  end
end

File: src/log_migration/authoring/repository.py
import os
import uuid
from dataclasses import dataclass, replace
from datetime import datetime
from pathlib import Path
from typing import Callable
from zoneinfo import ZoneInfo

from .database import AuthoringDatabase
from .frontmatter import parse_document, serialize_document
from .links import extract_wiki_links, replace_wiki_links
from .models import ConflictError, PageDocument, PageProblem, Redirect, SaveRequest, Status
from .names import page_path, validate_page_name


TOKYO = ZoneInfo("Asia/Tokyo")


@dataclass(frozen=True)
class RepositorySnapshot:
    pages: tuple[PageDocument, ...] = ()
    problems: tuple[PageProblem, ...] = ()
    redirects: tuple[Redirect, ...] = ()

    def with_redirect(self, old_route: str, new_route: str) -> "RepositorySnapshot":
        redirect = Redirect(old_route=old_route, new_route=new_route)
        if redirect in self.redirects:
            return self
        return replace(self, redirects=(*self.redirects, redirect))


class FileTransaction:
    def __init__(self) -> None:
        self._writes: dict[Path, bytes] = {}
        self._deletes: set[Path] = set()

    def write(self, path: Path, content: bytes) -> None:
        self._writes[path] = content
        self._deletes.discard(path)

    def delete(self, path: Path) -> None:
        if path not in self._writes:
            self._deletes.add(path)

    def commit(self) -> None:
        affected = set(self._writes) | self._deletes
        original = {path: path.read_bytes() for path in affected if path.exists()}
        created = {path for path in affected if path not in original}
        try:
            for path, content in self._writes.items():
                self._replace(path, content)
            for path in self._deletes:
                path.unlink()
        except Exception:
            for path, content in original.items():
                self._replace(path, content)
            for path in created:
                if path.exists():
                    path.unlink()
            raise

    @staticmethod
    def _replace(path: Path, content: bytes) -> None:
        path.parent.mkdir(parents=True, exist_ok=True)
        temporary = path.with_name(f".{path.name}.{uuid.uuid4().hex}.tmp")
        try:
            with temporary.open("xb") as stream:
                stream.write(content)
                stream.flush()
                os.fsync(stream.fileno())
            os.replace(temporary, path)
        finally:
            if temporary.exists():
                temporary.unlink()


class ContentRepository:
    def __init__(
        self,
        content_dir: Path,
        database_path: Path,
        clock: Callable[[], datetime],
    ):
        self.content_dir = content_dir
        self.database_path = database_path
        self.clock = clock
        self.database = AuthoringDatabase(database_path)

    def refresh(self) -> RepositorySnapshot:
        pages: list[PageDocument] = []
        problems: list[PageProblem] = []
        ids: set[str] = set()
        routes: set[str] = set()
        if self.content_dir.exists():
            for path in sorted(self.content_dir.glob("*.md")):
                try:
                    text = path.read_text(encoding="utf-8")
                except UnicodeDecodeError as error:
                    problems.append(PageProblem(path, f"document is not UTF-8: {error}"))
                    continue
                parsed = parse_document(path, text)
                if isinstance(parsed, PageProblem):
                    problems.append(parsed)
                    continue
                document = self._with_links_and_tokyo_time(parsed)
                if document.id in ids:
                    problems.append(PageProblem(path, f"duplicate page id: {document.id}"))
                    continue
                if document.route in routes:
                    problems.append(PageProblem(path, f"duplicate route: {document.route}"))
                    continue
                ids.add(document.id)
                routes.add(document.route)
                pages.append(document)
        snapshot = RepositorySnapshot(tuple(pages), tuple(problems))
        self.database.rebuild(snapshot)
        return snapshot

    def get_page(self, page_id: str) -> PageDocument | None:
        return next((page for page in self.refresh().pages if page.id == page_id), None)

    def find_route(self, route: str) -> PageDocument | None:
        normalized = route.strip("/")
        return next((page for page in self.refresh().pages if page.route == normalized), None)

    def list_pages(
        self,
        query: str = "",
        status: Status | None = None,
        empty_only: bool = False,
    ) -> tuple[PageDocument, ...]:
        pages = self.refresh().pages
        query = query.casefold()
        return tuple(
            page
            for page in pages
            if (status is None or page.status == status)
            and (not empty_only or page.is_empty)
            and (
                not query
                or query in page.display_title.casefold()
                or query in page.route.casefold()
            )
        )

    def save_draft(self, request: SaveRequest) -> PageDocument:
        snapshot = self.refresh()
        current = self._current_page(snapshot, request.page_id)
        if current is None:
            document = self._new_document(snapshot, request)
        else:
            self._validate_expected_update(current, request)
            document = replace(
                current,
                body=request.body,
                title=request.title if current.page_type == "date" else None,
                updated_at=self._now(),
            )
        document = self._with_links_and_tokyo_time(document)
        transaction = FileTransaction()
        transaction.write(document.path, serialize_document(document).encode())
        current_names = {page.name for page in snapshot.pages if page.name is not None}
        for link in document.links:
            name = validate_page_name(link.name)
            if name not in current_names:
                empty = self._new_empty_named_page(snapshot, name)
                transaction.write(empty.path, serialize_document(empty).encode())
                current_names.add(name)
        transaction.commit()
        self.refresh()
        return self.get_page(document.id) or document

    def rename_named_page(self, page_id: str, new_name: str) -> PageDocument:
        snapshot = self.refresh()
        page = self._current_page(snapshot, page_id)
        if page is None:
            raise ConflictError(f"page does not exist: {page_id}")
        if page.page_type != "named":
            raise ConflictError("only named pages can be renamed")
        new_name = validate_page_name(new_name)
        if new_name == page.name:
            return page
        destination = page_path(self.content_dir, "named", new_name, None)
        if any(other.name == new_name for other in snapshot.pages) or destination.exists():
            raise ConflictError(f"page name already exists: {new_name}")
        renamed = replace(page, name=new_name, path=destination, updated_at=self._now())
        transaction = FileTransaction()
        for source in snapshot.pages:
            if source.id == page.id:
                transaction.write(destination, serialize_document(renamed).encode())
            else:
                body = replace_wiki_links(source.body, page.name or "", new_name)
                if body != source.body:
                    transaction.write(
                        source.path,
                        serialize_document(replace(source, body=body, updated_at=self._now())).encode(),
                    )
        transaction.delete(page.path)
        transaction.commit()
        self.refresh()
        return self.get_page(page.id) or renamed

    def _new_document(self, snapshot: RepositorySnapshot, request: SaveRequest) -> PageDocument:
        now = self._now()
        if request.page_type == "date":
            if request.page_date is None:
                raise ValueError("date pages require page_date")
            if any(page.page_date == request.page_date for page in snapshot.pages):
                raise ConflictError(f"date page already exists: {request.page_date.isoformat()}")
            path = page_path(self.content_dir, "date", None, request.page_date)
            if path.exists():
                raise ConflictError(f"source path already exists: {path.name}")
            return PageDocument(
                id=uuid.uuid4().hex,
                page_type="date",
                name=None,
                page_date=request.page_date,
                title=request.title,
                status="draft",
                created_at=now,
                updated_at=now,
                published_at=None,
                path=path,
                body=request.body,
            )
        if request.page_type == "named":
            if request.name is None:
                raise ValueError("named pages require a name")
            name = validate_page_name(request.name)
            if any(page.name == name for page in snapshot.pages):
                raise ConflictError(f"page name already exists: {name}")
            path = page_path(self.content_dir, "named", name, None)
            if path.exists():
                raise ConflictError(f"source path already exists: {path.name}")
            return PageDocument(
                id=uuid.uuid4().hex,
                page_type="named",
                name=name,
                page_date=None,
                title=None,
                status="draft",
                created_at=now,
                updated_at=now,
                published_at=None,
                path=path,
                body=request.body,
            )
        raise ValueError(f"unknown page type: {request.page_type}")

    def _new_empty_named_page(self, snapshot: RepositorySnapshot, name: str) -> PageDocument:
        return self._new_document(snapshot, SaveRequest(page_type="named", name=name, body=""))

    @staticmethod
    def _current_page(snapshot: RepositorySnapshot, page_id: str | None) -> PageDocument | None:
        if page_id is None:
            return None
        return next((page for page in snapshot.pages if page.id == page_id), None)

    @staticmethod
    def _validate_expected_update(page: PageDocument, request: SaveRequest) -> None:
        if request.expected_updated_at is not None and request.expected_updated_at != page.updated_at:
            raise ConflictError("page was updated by another edit")

    def _now(self) -> datetime:
        value = self.clock()
        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("clock must return an aware datetime")
        return value.astimezone(TOKYO)

    @staticmethod
    def _with_links_and_tokyo_time(document: PageDocument) -> PageDocument:
        times = (document.created_at, document.updated_at, document.published_at)
        if any(value is not None and (value.tzinfo is None or value.utcoffset() is None) for value in times):
            raise ValueError("document timestamps must be aware")
        return replace(
            document,
            created_at=document.created_at.astimezone(TOKYO),
            updated_at=document.updated_at.astimezone(TOKYO),
            published_at=document.published_at.astimezone(TOKYO) if document.published_at else None,
            links=extract_wiki_links(document.body),
        )
